### ISSUE TITLE
Sensitive species 2: additional columns for download

### ISSUE THREAD

**Issue by kitenetter** [BiologicalRecordsCentre/iRecord#1714]
Add new functionality to the csv download process for iRecord, for both ‘standard’ and ‘simple’ downloads:
- [x] Retain the blurred grid ref in the output grid ref column
- [x] The location name should default to “<sensitive record, location hidden>” (using “<>” prefix/suffix so that this location can can be clearly distinguished from on-sensitive sites when ordering alphabetically by location name; could use a different prefix/suffix if preferable, e.g. “*** sensitive record, location hidden***”)
- [x] Add a new column “sensitive record”, set to TRUE if the record is flagged as sensitive due to taxon or user-specified sensitivity, otherwise FALSE
- [x] Add a new column “sensitive location” for the entered location name, only populated if the record is sensitive
- [x] Add a new column “sensitive output grid ref” for the unblurred grid ref, only populated if the record is sensitive
- [ ] Add a new column “sensitive VC” for the VC based on the precise location, only populated if the record is sensitive

To enable the above, we will need to add indexing based on the true location, as well as on the blurred.

ES data already supports a blurred and an unblurred view, will need to add a database field to enable this in Postgres.

**Comment by kitenetter** [comment]
Question: once the above is available in the standard download, will it automatically feed through to the survey-specific downloads, or will they need additional editing?

**Comment by johnvanbreda** [comment]
> once the above is available in the standard download, will it automatically feed through to the survey-specific downloads, or will they need additional editing?

Yes, the survey specific downloads are the same as the standard downloads, with the addition of extra columns for custom attributes.

Some questions:
1. In the standard download, there is an existing location field called site name, so shall I call the new field sensitive site rather than sensitive location? 
2. On the simple download, there is a Site field which has some special handling - replacing the value with 'unnamed site' if the location name is empty. This is one of the adaptations made so that the Simple download format would be MapMate compatible. Presumably we can alter this field and populate the value with 'sensitive record, location hidden' if the record is sensitive without breaking MapMate imports?
3. Does the sensitive location/site field get populated for all users, or only verifiers (and people with sensitive data access)?

**Comment by kitenetter** [comment]
1. Yes, let's stick to using site
2. Yes, 'sensitive record, location hidden' should work for MapMate, but note request to have a non alphabetic wrapper around the name
3. The sensitive data should be available to users downloading their own records (= anyone using 'My records' from the 'Records to access' list); to verifiers (i.e. anyone with a verification role and/or using a verification filter from the 'Records to access' list); and to LERCs (the only definition we have for these is the use of a download filter that is prefixed "LERC" in the 'Records to access' list). The difficulty is that we have other custom downloads set up for which we do not want sensitive details to be exposed, or at least not without checking first. Happy to go back to giving LERC accounts a defined role if that will make this easier to set up.

**Comment by johnvanbreda** [comment]
More questions:
1. When you don't have sensitive records access, the system blocks access to the site name at a low level, so the code can't currently know whether there is a site name or not. Therefore we can't state that the site name is being hidden as it might just not exist at all (and on the Simple download version we can't differentiate between "unnamed site" and "[sensitive record, location hidden]"). One possible solution would be to store ! (or some other similar indicator) to indicate a site name being withheld because of sensitivity and store an empty string if there was no site name, so we can differentiate. Would that be OK? This would affect the outputs on normal explore pages - the site name for a sensitive record would show ! if there is a withheld site name.
2. For the LERC download, to avoid making the permissions in Indicia yet more complex I would suggest we have a separate clone of the Download page which you give access to using an LERC role in Drupal. This would show their downloads (perhaps we need to filter to those starting LERC...) and use a custom Elasticsearch connection so that it gets the full precision version of each record. Is that OK?

**Comment by kitenetter** [comment]
1. Main concern over using "!" is that it isn't obvious what it means (e.g. to someone visiting the Explore page). But it sounds like we don't have a lot of options here, and the use of "!" does at least match the red symbol that we already use for sensitive records, so this seems workable.
2. I think that is okay. We have a role "regional collator" that doesn't currently have a function - as far as I can see that role does not grant any permissions beyond those of a standard user, so we could go back to using that role for the LERCs. But maybe it would be clearer to create a new role (or rename that old one).

All downloads for LERCs are prefixed "LERC" so that will work for filtering.

**Comment by johnvanbreda** [comment]
I've done all the coding required for this. Once deployed, I'll need to set up the special version of the download page with custom ES alias for LERCs.

**Comment by kitenetter** [comment]
@johnvanbreda let me know if you need me to start assigning a new role to the LERC user accounts.

**Comment by kitenetter** [comment]
@johnvanbreda I'm just testing this, based on a verifier download for some sensitive records.

**Working as expected:**
- We have a column for "Site name" that is populated with "[sensitive record, location hidden]"
- We have a column for "Sensitive site" that is populated with the original site name

**Not working as expected:**
- Column "Output grid ref" contains the full-precision grid ref - original spec was to retain the blurred grid ref in the "Output grid ref" column, and add a column for "Sensitive output grid ref"
- Likewise we had suggested retaining a VC based on the blurred grid ref in the VC columns, and adding new columns for VC based on the sensitive grid ref
- We had specified adding a new column “sensitive record”, set to TRUE if the record is flagged as sensitive due to taxon or user-specified sensitivity, otherwise FALSE

Happy to review options for the above if needed, but it is important that we retain separation of the blurred location fields (which could be used for wider data sharing) and the unblurred sensitive location fields.

**Comment by johnvanbreda** [comment]
> * Column "Output grid ref" contains the full-precision grid ref - original spec was to retain the blurred grid ref in the "Output grid ref" column, and add a column for "Sensitive output grid ref"
>  * Likewise we had suggested retaining a VC based on the blurred grid ref in the VC columns, and adding new columns for VC based on the sensitive grid ref

This would require additional fields in Elasticsearch, a few hours work plus a re-population of all the sensitive records in ES. Let me know if this is a priority.

I've added code for the sensitive flag so it will be available after the next release of warehouse code.

**Comment by kitenetter** [comment]
We do need to complete the work needed to implement the sensitive species blurring, so this needs to remain in the current priorities.

**Comment by johnvanbreda** [comment]
In the develop branch, there is code now so that the output grid ref will be blurred and there is a sensitive output grid ref column.

**Comment by johnvanbreda** [comment]
Blurring the output grid ref, and adding a sensitive output grid ref, is now live.

**Comment by johnvanbreda** [comment]
@kitenetter one thing that has come to light when implementing the last part of this is that when a record is blurred to 100km (or when a record is simply low precision) the spatial indexing links that record and sample to a huge list of locations, e.g. RSPB reserves, NUTS level 3, National Trust Sites etc. These are taking quite a bit of disk space on Elasticsearch and the problem is made worse now that we store 2 copies of the indexed locations. With this change, the verification filtering is done on an accurate list of indexed vice counties so the blurred version is only partially useful (e.g. it gets used when filtering by a location on the Explore page). If you filter to an RSPB reserve you might argue that 100km blurred records should be excluded anyway, even if there is a tiny chance they are from the reserve.
I propose that the behaviour is changed for 100km record indexing, so that they are only indexed at the country level. This will significantly reduce disk space required for these records and reduce "false positives" when doing spatially indexed filtering on the Explore page and should not affect verification at all. Are you happy with this proposal? If you like, we could include the LRC boundary and Vice County level in the indexing so that these records are included if a VC/LRC boundary filter is applied on the Explore page.

**Comment by kitenetter** [comment]
@johnvanbreda I'm happy to exclude the 100km indexing and reduce false positives.

**Comment by johnvanbreda** [comment]
> Add a new column “sensitive VC” for the VC based on the precise location, only populated if the record is sensitive

@kitenetter Given that the VC should now be based on the full precision record if you have full precision access to the record (i.e. for verifiers), do we need to implement this?

Presumably it would mean switching the normal VC output columns to the blurred version and adding VC sensitive output columns for the full precision versions. Other than changing the download configuration in the REST API I think everything else is in place for this.

**Comment by kitenetter** [comment]
@johnvanbreda I don't understand your question in the comment immediately above, let me know if we need to discuss.

**Comment by johnvanbreda** [comment]
@kitenetter my point was that previously, a sensitive record's Vice County was always indexed against the blurred copy of the record. This is still the case for the public view a record (so a 100km blur may result in several VCs being allocated), but we have made changes elsewhere so that the full precision copy of the record which is available to verifiers has the VC allocation built using the full precision copy of the record. So the default VC output columns are accurate for verifiers and I'm not sure there is a need to have columns for both the blurred indexing results and the full precision indexing results.

**Comment by kitenetter** [comment]
@johnvanbreda if "the default VC output columns are accurate for verifiers" then I agree we don't need to provide an additional VC column for verifier downloads.
